### PR TITLE
PeerPrep Project Completion

### PR DESCRIPTION
**Pull request by Jai2501**
- Add various Microservices and Frontend
- Dockerized Microservices
- Performed Testing
- Add README.md

### CODE REVIEW

**Review by Nanxi-Huang — COMMENTED** [review]
LGTM👍

File: HistoryService/Controllers/HistoryDatabaseController.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  );
+
+  return result;
+}
+
+module.exports = {
+  connectToDatabase,
+  getAttemptDetailsFromDatabase,
+  addAttemptDetailsToDatabase,
+};
```
Comment (Nanxi-Huang) [inline HistoryService/Controllers/HistoryDatabaseController.js:136]: very good code style!

File: HistoryService/Controllers/HistoryRoutesController.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    res.status(500).json({
+      message: "HISTORY DATA COULD NOT BE ADDED",
+      error: "Missing Fields or Internal Error",
+    });
+  }
+};
+
+module.exports = {
+  getAttemptDetails,
+  addAttemptDetails,
```
Comment (Nanxi-Huang) [inline HistoryService/Controllers/HistoryRoutesController.js:77]: good code style!

**Review by Nanxi-Huang — DISMISSED** [review]

File: Frontend/src/Components/Collaboration/CodeEditor.jsx
Comment (Jai2501) [inline Frontend/src/Components/Collaboration/CodeEditor.jsx:1]: Good job on the code editor.

File: Frontend/src/Components/Collaboration/CommunicationWindow.jsx [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        peerConnection.current.addTrack(track, localStream);
+      });
+
+      const offer = await peerConnection.current.createOffer();
+      await peerConnection.current.setLocalDescription(offer);
+      socket.emit("call", offer);
+      // Call is now waiting for the partner to accept
+      showToast("Waiting for partner to join the call...");
+    } catch (err) {
+      showToast("Failed to access media devices.");
```
Comment (Jai2501) [inline Frontend/src/Components/Collaboration/CommunicationWindow.jsx:126]: Good use of Toasts, informing the user of whats happening.

File: Frontend/src/Components/GenerativeAI/AIChat.jsx
Comment (Jai2501) [inline Frontend/src/Components/GenerativeAI/AIChat.jsx:1]: Well implemented.

File: Frontend/src/Components/GenerativeAI/AIService.jsx
Comment (Jai2501) [inline Frontend/src/Components/GenerativeAI/AIService.jsx:1]: Concise and to the point!

File: Frontend/src/Components/LandingPage/AttemptView.jsx
Comment (Jai2501) [inline Frontend/src/Components/LandingPage/AttemptView.jsx:1]: Well implemented!

File: Frontend/src/Components/LandingPage/UserQuestionList.jsx
Comment (Jai2501) [inline Frontend/src/Components/LandingPage/UserQuestionList.jsx:1]: Clean Implementation

File: Frontend/src/Components/Matching/NotSuccessOutput.jsx
Comment (Jai2501) [inline Frontend/src/Components/Matching/NotSuccessOutput.jsx:1]: Well thought of!

File: Frontend/src/Components/QuestionList/QLanguageDropdown.jsx
Comment (Jai2501) [inline Frontend/src/Components/QuestionList/QLanguageDropdown.jsx:1]: The separation of UI components is good and improves modularity.

File: MatchingService/controllers/matchingController.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                await res.status(500).json(matchResult);
+                break;
+
+            case 'cancel':
+                await res.status(200).json(matchResult);
+                break;
+
+            default:
+                await res.status(500).json({ message: 'Unknown error. Please try again!' });
+        }
```
Comment (Jai2501) [inline MatchingService/controllers/matchingController.js:31]: Good to handle status errors in this way

File: MatchingService/database/matchedPairDb.js
Comment (Jai2501) [inline MatchingService/database/matchedPairDb.js:1]: Good Segregation of Database and Route Controllers

File: MatchingService/test/matchingService.test.js
Comment (Jai2501) [inline MatchingService/test/matchingService.test.js:1]: Good to test!

File: QuestionService/models/questionModel.js
Comment (Jai2501) [inline QuestionService/models/questionModel.js:1]: Question Schema well defined

File: QuestionService/populateDb.js
Comment (Jai2501) [inline QuestionService/populateDb.js:1]: Good to have some initial values on startup.

**Review by Jai2501 — COMMENTED** [review]
LGTM!

All have done a fantastic job!

File: Frontend/src/Components/Matching/MatchButton.jsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+//    const [chosenLanguage, setChosenLanguage] = useState("No Preference");
+//    const [chosenProficiency, setChosenProficiency] = useState("No Preference");
+//    const [chosenTopic, setChosenTopic] = useState("No Preference");
+
+    const handleMatchInitiation = (matchCriteria) => {
+        setGoToLoadPopup(true);
+        setViewPopup(false);
+
+        axios.post(`/api/match`, matchCriteria) // replace /${props.userId} with how we access userId
+        .then(response => {
```
Comment (laihuiqi) [inline Frontend/src/Components/Matching/MatchButton.jsx:25]: Good handling of response!

**Review by laihuiqi — APPROVED** [review]
Overall, LGTM.

**Review by Nanxi-Huang — APPROVED** [review]
LGTM!

**Pull request merged**